# **Appendix A: Introduction to PyTorch (Part 1)**

## **1. What is PyTorch**

### **1.1 The three core components of PyTorch**

As shown on **Fig. 1**, PyTorch is built upon three core pillars designed to streamline the transition from research to production:

* **Tensor Library:** An extension of NumPy that enables seamless switching between CPU and GPU for hardware-accelerated computation.
* **Autograd Engine:** An automatic differentiation system that handles gradient computations, simplifying the backpropagation process essential for training models.
* **Deep Learning Framework:** A modular ecosystem providing high-level building blocks—such as pre-trained models, loss functions, and optimizers—to efficiently design and scale neural networks.

The upcoming sections will detail these components through hands-on code examples.

<figure style='text-align:center'>
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/1.webp" width="800px">
<figcaption><b>Figure 1.</b> PyTorch's three main components include a tensor library as a fundamental building block for computing, automatic differentiation for model optimization, and deep learning utility functions, making it easier to implement and train deep neural network models.</figcaption>
</figure>

### **1.2 Defining deep learning**

Deep learning is a subcategory of machine learning that is focused on the implementation of deep neural networks. In turn, machine learning is a subcategory of AI that is concerned with algorithms that learn from data. AI is the broader concept of machines being able to perform tasks that typically require human intelligence.

<figure style='text-align: center'>
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/2.webp" width="600px">
</figure>

The supervised learning workflow for predictive modeling consists of a training stage where a model is trained on labeled examples in a training dataset. The trained model can then be used to predict the labels of new observations.

<figure style='text-align: center'>
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/3.webp" width="600px">
</figure>

### **1.3 Installing PyTorch**

PyTorch installation involves selecting the right version for your hardware and ensuring environment compatibility. Here is a concise breakdown:

#### **1.3.1 Environment & Python Version**

Use a Python version 1–2 releases behind the latest (e.g., if Python 3.13 is out, use 3.11 or 3.12) to ensure library support. Confirm your installation by running `import torch; print(torch.__version__)`.

In [1]:
import torch

print(torch.__version__)

2.10.0


### **1.3.2 Hardware Acceleration**

* **NVIDIA (CUDA):** The standard `pip install torch` usually detects CUDA, but visiting [pytorch.org](https://pytorch.org) is recommended to get the specific command for your OS and CUDA version. Run `torch.cuda.is_available()` to confirm PyTorch recognizes your NVIDIA GPU.

<figure style='text-align: center'>
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/4.webp" width="600px">
</figure>

* **AMD (ROCm):** Supported via experimental builds; check the official site for specific instructions.
* **Apple Silicon (MPS):** Uses the Metal Performance Shaders (MPS) backend. Verify support with `torch.backends.mps.is_available()`.
* **Cloud Options:** If you lack a local GPU, use *Google Colab* to access free or paid GPU runtimes.

<figure style='text-align: center'>
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/5.webp" width="600px">
</figure>

>**Fun note:** The package is imported as `torch` (not `pytorch`) to honor its predecessor, the original Lua-based *Torch* scientific computing framework.

In [2]:
print(torch.cuda.is_available())

False


## **2. Understanding tensors**

Tensors are mathematical objects that generalize scalars, vectors, and matrices into multi-dimensional data containers. Their complexity is defined by their rank (or order), which represents the number of dimensions they possess (e.g., a rank-0 tensor is a scalar, while a rank-2 tensor is a matrix).

<figure style='text-align: center'>
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/6.webp" width="600px">
</figure>

Key characteristics of tensors are as follow:
* **Data Containers:** Tensors function as multi-dimensional arrays used to store and manipulate features efficiently.
* **Compared to NumPy:** While PyTorch adopts much of the NumPy API and syntax, it extends these capabilities with two critical features for deep learning:
    * **Automatic Differentiation:** Built-in gradient computation (Autograd).
    * **Hardware Acceleration:** Support for high-speed training on GPUs.

The following sections will explore creating and manipulating these tensors using PyTorch’s NumPy-like interface.

### **2.1 Scalars, vectors, matrices, and tensors**

As mentioned earlier, PyTorch tensors are data containers for array-like structures. A scalar is a 0-dimensional tensor (for instance, just a number), a vector is a 1-dimensional tensor, and a matrix is a 2-dimensional tensor. There is no specific term for higher-dimensional tensors, so we typically refer to a 3-dimensional tensor as just a 3D tensor, and so forth.

We can create objects of PyTorch’s Tensor class using the torch.tensor function as follows:

In [3]:
import numpy as np

# create a 0D tensor (scalar) from a Python integer
tensor0d = torch.tensor(1)

# create a 1D tensor (vector) from a Python list
tensor1d = torch.tensor([1, 2, 3])

# create a 2D tensor from a nested Python list
tensor2d = torch.tensor([[1, 2], 
                         [3, 4]])

# create a 3D tensor from a nested Python list
tensor3d_1 = torch.tensor([[[1, 2], [3, 4]], 
                           [[5, 6], [7, 8]]])

# create a 3D tensor from NumPy array
ary3d = np.array([[[1, 2], [3, 4]], 
                  [[5, 6], [7, 8]]])
tensor3d_2 = torch.tensor(ary3d)  # Copies NumPy array
tensor3d_3 = torch.from_numpy(ary3d)  # Shares memory with NumPy array

In [4]:
ary3d[0, 0, 0] = 999
print(tensor3d_2) # remains unchanged

tensor([[[1, 2],
         [3, 4]],

        [[5, 6],
         [7, 8]]])


In [5]:
print(tensor3d_3) # changes because of memory sharing

tensor([[[999,   2],
         [  3,   4]],

        [[  5,   6],
         [  7,   8]]])


### **2.2 Tensor data types**

PyTorch automatically assigns data types based on the input values, prioritizing a balance between numerical precision and hardware performance. Here are default data types:
* **Integers:** Created from Python integers, defaulting to 64-bit (`torch.int64`).
* **Floats:** Created from Python floats, defaulting to 32-bit (`torch.float32`).
* *Note:* 32-bit floats are the deep learning standard because they consume less memory and are heavily optimized for GPU acceleration.

In [6]:
print(tensor1d)
print(tensor1d.dtype)

tensor([1, 2, 3])
torch.int64


In [7]:
floatvec = torch.tensor([1.0, 2.0, 3.0])
print(floatvec.dtype)

torch.float32


You can check a tensor's type using the `.dtype` attribute and modify it using the `.to()` method:

In [8]:
floatvec = tensor1d.to(torch.float32)
print(floatvec)
print(floatvec.dtype)

tensor([1., 2., 3.])
torch.float32


### **2.3 Common PyTorch tensor operations**

While PyTorch offers a vast library of operations, several essential commands form the foundation of most deep learning projects. Here are some of core tensor operations in PyTorch:
* **Shape & Inspection:** Use `.shape` to check dimensions (e.g., rows and columns).
* **Reshaping:** Modify a tensor's structure using `.view()` (the PyTorch standard) or `.reshape()`.
* **Transposition:** Use `.T` to flip a tensor across its diagonal, effectively swapping its axes.
* **Matrix Multiplication:** Use the `.matmul()` method or the compact **`@`** operator for dot products.

In [9]:
tensor2d = torch.tensor([[1, 2, 3], 
                         [4, 5, 6]])
print(tensor2d)
print(tensor2d.shape)

tensor([[1, 2, 3],
        [4, 5, 6]])
torch.Size([2, 3])


In [10]:
print(tensor2d.reshape(3,2))
print(tensor2d.view(3,2))
print(tensor2d.T)

tensor([[1, 2],
        [3, 4],
        [5, 6]])
tensor([[1, 2],
        [3, 4],
        [5, 6]])
tensor([[1, 4],
        [2, 5],
        [3, 6]])


In [11]:
print(tensor2d.matmul(tensor2d.T))
print(tensor2d @ tensor2d.T)

tensor([[14, 32],
        [32, 77]])
tensor([[14, 32],
        [32, 77]])


## **3. Seeing models as computation graphs**

Following the tensor library, the second pillar of PyTorch is *Autograd*, an automatic differentiation engine that operates using computational graphs.

The Computational Graph is a directed graph that visualizes and organizes mathematical expressions as a sequence of operations. It maps the "forward pass" (prediction step) of a model, tracking how inputs and parameters flow through functions to produce an output or loss. PyTorch builds this graph in the background to enable backpropagation, the process of calculating gradients required to optimize model parameters.

### **Example: Logistic Regression**

In a simple logistic regression classifier, the graph connects input features ($x$), weights ($w$), and biases ($b$) to compute an activation ($a$) and a final loss. This structured sequence allows PyTorch to automatically determine how changing a weight affects the total error.

<figure style='text-align: center'>
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/7.webp" width="800px">
</figure>

In [12]:
import torch.nn.functional as F

y = torch.tensor([1.0])  # true label
x1 = torch.tensor([1.1]) # input feature
w1 = torch.tensor([2.2]) # weight parameter
b = torch.tensor([0.0])  # bias unit

z = x1 * w1 + b          # net input
a = torch.sigmoid(z)     # activation & output

loss = F.binary_cross_entropy(a, y)
print(loss)

tensor(0.0852)


## **4 Automatic differentiation made easy**

PyTorch’s Autograd engine automates the complex calculus required to train neural networks, specifically by calculating gradients through a process called backpropagation.

<figure style='text-align: center'>
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/8.webp" width="800px">
</figure>

By setting `requires_grad=True` on a tensor (like weights or biases), you tell PyTorch to track all operations involving that tensor to build a computation graph. PyTorch uses *reverse-mode automatic differentiation* to apply the chain rule from the loss back to the input, determining how each parameter should be updated to minimize error. To run backpropagation manually, we run `grad(objective, param)`. By default, PyTorch destroys the computation graph after calculating the gradients to free memory. However, since we are going to reuse this computation graph with `.backward()` method, we set `retain_graph=True` inside `grad()` function so that it stays in memory.

Instead of manual derivation, calling `.backward()` on a loss tensor automatically calculates gradients for all relevant parameters and stores them in their respective `.grad` attributes for all leaves of the computational graph. While you can use the `grad()` function for manual experimentation, the standard workflow in PyTorch leverages the `.backward()` method to handle the math automatically.

> **Takeaway:** You don't need to perform manual calculus; PyTorch tracks the operations and computes the derivatives for you with a single method call.

In [13]:
import torch.nn.functional as F
from torch.autograd import grad

y = torch.tensor([1.0])
x1 = torch.tensor([1.1])
w1 = torch.tensor([2.2], requires_grad=True)
b = torch.tensor([0.0], requires_grad=True)

z = x1 * w1 + b 
a = torch.sigmoid(z)

loss = F.binary_cross_entropy(a, y)

grad_L_w1 = grad(loss, w1, retain_graph=True)
grad_L_b = grad(loss, b, retain_graph=True)

print(grad_L_w1)
print(grad_L_b)

(tensor([-0.0898]),)
(tensor([-0.0817]),)


In [14]:
loss.backward()

print(w1.grad)
print(b.grad)

tensor([-0.0898])
tensor([-0.0817])


## **5. Implementing multilayer neural networks**

The third pillar of PyTorch is its Deep Learning Library, centered around the `torch.nn.Module` class, which provides the framework for building complex architectures like Multilayer Perceptrons (MLPs).

<figure style='text-align: center'>
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/9.webp" width="600px">
</figure>

* Subclassing `torch.nn.Module` allows you to bundle layers, operations, and parameters into a single, manageable object. The `torch.nn.Module` class (the parent) contains complex internal logic for tracking gradients, registering parameters, and handling GPU/CPU moves. If you don't call `super().__init__()`, your child class is just an empty shell.
* The Constructor (`__init__`) is used to define the structural building blocks of the network, such as fully connected (linear) layers or activation functions.
* The forward pass (`forward`) defines the data flow from input to output. When you call this method, PyTorch automatically constructs the computation graph required for training.
* When you define the flow in the `forward` method, PyTorch’s autograd engine automatically handles the `backward` pass (gradient calculation) without requiring manual coding.

When implementing a network, you focus on defining **what** the layers are and **how** data moves through them; the library handles the underlying parameter tracking and calculus.

In [15]:
class NeuralNetwork(torch.nn.Module):
    def __init__(self, num_inputs, num_outputs):
        super().__init__()

        self.layers = torch.nn.Sequential(
                
            # 1st hidden layer
            torch.nn.Linear(num_inputs, 30),
            torch.nn.ReLU(),

            # 2nd hidden layer
            torch.nn.Linear(30, 20),
            torch.nn.ReLU(),

            # output layer
            torch.nn.Linear(20, num_outputs),
        )

    def forward(self, x):
        logits = self.layers(x)
        return logits

We can then instantiate a new neural network object. It is often useful to call print on the model to see a summary of its structure:

In [16]:
model = NeuralNetwork(50, 3)
print(model)

NeuralNetwork(
  (layers): Sequential(
    (0): Linear(in_features=50, out_features=30, bias=True)
    (1): ReLU()
    (2): Linear(in_features=30, out_features=20, bias=True)
    (3): ReLU()
    (4): Linear(in_features=20, out_features=3, bias=True)
  )
)


The `torch.nn.Sequential` container is a productivity tool used to group layers that execute in a fixed, linear order. Here are key advantages of sequenntial layer implementation:

* **Code Simplicity:** Instead of calling each layer individually (e.g., `x = self.layer1(x)`, `x = self.relu(x)`, etc.) inside the `forward` method, you simply call the `Sequential` object once.
* **Organization:** It consolidates the architecture within the `__init__` constructor, making the network's structure easier to read and maintain.

While convenient for linear architectures, `Sequential` is less suitable for complex models requiring non-linear data flows, such as skip connections or multiple inputs/outputs. We can rewrite `NeuralNetwork` class using the functional API as below:

In [17]:
import torch.nn.functional as F

class NeuralNetworkFunctional(torch.nn.Module):
    def __init__(self, num_inputs, num_outputs):
        super().__init__()

        # Instantiate functional calls for three layers    
        self.fc1 = torch.nn.Linear(num_inputs, 30)
        self.fc2 = torch.nn.Linear(30, 20)
        self.fc3 = torch.nn.Linear(20, num_outputs)
            
    def forward(self, x):
        
        # 1st layer
        x = self.fc1(x)
        x = F.relu(x)
        
        # 2nd layer
        x = self.fc2(x)
        x = F.relu(x)
        
        # Output logits
        logits = self.fc3(x)         
        return logits

In [21]:
# Create dummy input data (Batch Size = 5, Features = 50)
x = torch.rand((10, 50))
print(x)

tensor([[0.2391, 0.3194, 0.8111, 0.7507, 0.3306, 0.5374, 0.2845, 0.8459, 0.2232,
         0.2083, 0.8169, 0.1084, 0.3285, 0.7185, 0.3624, 0.3084, 0.8893, 0.4179,
         0.9741, 0.3697, 0.2397, 0.8936, 0.1443, 0.1365, 0.7625, 0.1632, 0.6641,
         0.1525, 0.9830, 0.5936, 0.9120, 0.0146, 0.6323, 0.4743, 0.7467, 0.3545,
         0.9994, 0.9815, 0.7399, 0.2057, 0.8742, 0.0138, 0.7676, 0.7481, 0.7570,
         0.6432, 0.9111, 0.2246, 0.8668, 0.6961],
        [0.5131, 0.6978, 0.4537, 0.9035, 0.5088, 0.7786, 0.9455, 0.6622, 0.5138,
         0.4649, 0.0926, 0.1094, 0.6668, 0.5465, 0.6295, 0.0483, 0.7799, 0.4483,
         0.6947, 0.2243, 0.6045, 0.7574, 0.1262, 0.5446, 0.3269, 0.9105, 0.3953,
         0.2075, 0.1796, 0.4544, 0.7271, 0.6692, 0.9545, 0.8872, 0.5824, 0.6379,
         0.2836, 0.6754, 0.8838, 0.4898, 0.5963, 0.0890, 0.7804, 0.9223, 0.9605,
         0.7099, 0.3075, 0.5226, 0.2881, 0.2615],
        [0.9618, 0.5225, 0.1488, 0.4243, 0.7340, 0.4202, 0.8299, 0.7550, 0.0284,
         

In PyTorch, every time you instantiate a `torch.nn.Linear` layer, the weights and biases are initialized with small random numbers. Even if the code looks the same, the "starting point" for those numbers is different every time. To compare the outputs of the same models, we must set a "seed" for the random number generator before creating each model:

In [19]:
# Set seed for sequential model 
torch.manual_seed(123)
model_sequential = NeuralNetwork(50, 3)

# Set seed for functional model
torch.manual_seed(123)
model_functional = NeuralNetworkFunctional(50, 3)

assert torch.allclose(model_sequential(x), model_functional(x), atol=1e-6)

Next, let’s check the total number of trainable parameters of this model:

In [22]:
num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print("Total number of trainable model parameters:", num_params)

Total number of trainable model parameters: 2213


We could manually calculate number of trainable parameters too:
* First hidden layer: $30 \times 50 + 30$
* Second hidden layer: $20 \times 30 + 20$
* Output layer: $3 \times 20 + 3$

If we sum up above numbers, we'll get the same number as above (2213).

Let's look at these layers one by one:

In [23]:
print(model.layers)

Sequential(
  (0): Linear(in_features=50, out_features=30, bias=True)
  (1): ReLU()
  (2): Linear(in_features=30, out_features=20, bias=True)
  (3): ReLU()
  (4): Linear(in_features=20, out_features=3, bias=True)
)


In [24]:
# Dimension of first hidden layer
print(model.layers[0].weight.shape)
print(model.layers[0].bias.shape)

torch.Size([30, 50])
torch.Size([30])


In [25]:
# Dimension of second hidden layer
print(model.layers[2].weight.shape)
print(model.layers[2].bias.shape)

torch.Size([20, 30])
torch.Size([20])


In [26]:
# Dimension of output layer
print(model.layers[4].weight.shape)
print(model.layers[4].bias.shape)

torch.Size([3, 20])
torch.Size([3])


As briefly mentioned above, deep learning models use random weight initialization to "break symmetry," ensuring that different neurons learn unique features rather than performing identical calculations during training. If all weights started at the same value (e.g., zero), every neuron would produce the same output and receive the same gradient update, preventing the model from learning complex patterns. While randomness is essential for learning, it can make debugging difficult. You can make results predictable by using `torch.manual_seed()`, which forces the random number generator to produce the same "random" sequence every time.

In [27]:
torch.manual_seed(123)

out = model(x)
print(out)

tensor([[ 0.0013, -0.0786, -0.0798],
        [-0.0056, -0.0955, -0.0834],
        [-0.0174, -0.0639, -0.0593],
        [ 0.0209, -0.1067, -0.0603],
        [-0.0071, -0.0432, -0.0510],
        [ 0.0134, -0.0774, -0.0435],
        [ 0.0449, -0.0588, -0.0742],
        [ 0.0181, -0.0938, -0.0446],
        [ 0.0187, -0.0985, -0.0865],
        [ 0.0321, -0.0970, -0.0705]], grad_fn=<AddmmBackward0>)


When you call `model(x)`, PyTorch executes the forward pass, transforming input features into output scores (logits) through each layer of the network. The computational graph tracks inputs flow from the input layer through hidden layers to the output. In this process, each resulting tensor contains a `grad_fn` attribute (e.g., `<AddmmBackward0>`). This "footprint" identifies the mathematical operations used, like matrix multiplication (`mm`) and addition (`Add`), so PyTorch can retrace its steps during backpropagation.

During inference, however, tracking gradients is unnecessary and resource-intensive. Wrapping your code in the `torch.no_grad()` context manager disables the computational graph. This reduces: (1) memory consumption by not storing intermediate tensors, and (2) speeds up computation by skipping gradient-related overhead.

In [28]:
with torch.no_grad():
    out = model(x)
print(out)

tensor([[ 0.0013, -0.0786, -0.0798],
        [-0.0056, -0.0955, -0.0834],
        [-0.0174, -0.0639, -0.0593],
        [ 0.0209, -0.1067, -0.0603],
        [-0.0071, -0.0432, -0.0510],
        [ 0.0134, -0.0774, -0.0435],
        [ 0.0449, -0.0588, -0.0742],
        [ 0.0181, -0.0938, -0.0446],
        [ 0.0187, -0.0985, -0.0865],
        [ 0.0321, -0.0970, -0.0705]])


In PyTorch, it is standard practice for models to return **raw logits** (the direct outputs of the final linear layer) rather than probabilities. The main reason is numerical stability. Common PyTorch loss functions (like `CrossEntropyLoss`) internally combine a nonlinear activation (Softmax/Sigmoid) with the loss calculation. This fused approach is more mathematically stable and efficient. Another reason is flexibility during training and inference. During training, you feed raw logits directly into the loss function while during inference you must explicitly apply `torch.softmax()` to the output to get human-readable class-membership probabilities (which sum to 1).

In [29]:
with torch.no_grad():
    out = torch.softmax(model(x), dim=1)
print(out)

tensor([[0.3515, 0.3245, 0.3241],
        [0.3522, 0.3219, 0.3259],
        [0.3432, 0.3276, 0.3291],
        [0.3569, 0.3141, 0.3290],
        [0.3423, 0.3301, 0.3276],
        [0.3499, 0.3195, 0.3306],
        [0.3585, 0.3232, 0.3183],
        [0.3529, 0.3156, 0.3315],
        [0.3585, 0.3188, 0.3227],
        [0.3595, 0.3160, 0.3245]])


## **6. Setting up efficient data loaders**

Efficient data loading in PyTorch relies on two specialized classes that separate the storage of data (`Dataset` class) from the logic of accessing it (`DataLoader` class): 

* **The `Dataset` Class:** Acts as the container for your raw data. It defines how to load an individual record and its corresponding label (e.g., loading an image from a file path).
* **The `DataLoader` Class:** Acts as the engine. It handles how data is served to the model by managing:
    * **Batching:** Grouping individual records into chunks (batches).
    * **Shuffling:** Randomizing the order of data to prevent the model from learning patterns based on sequence.
    * **Parallelization:** Using multiple CPU cores to load data while the GPU is training.

<figure style='text-align: center'>
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/10.webp" width="800px">
</figure>

To prepare for training, you typically create the dataset (features and labels) and split it into two distinct sets:

1. **Training Set:** Used by the model to learn weights.
2. **Test Set:** Used to evaluate the model's performance on unseen data.

Let’s start by creating a simple toy dataset of five training examples with two features each. Accompanying the training examples, we also create a tensor containing the corresponding class labels: three examples belong to class 0, and two examples belong to class 1. In addition, we also make a test set consisting of two entries. The code to create this dataset is shown below.

In [30]:
X_train = torch.tensor([
    [-1.2, 3.1],
    [-0.9, 2.9],
    [-0.5, 2.6],
    [2.3, -1.1],
    [2.7, -1.5]
])

y_train = torch.tensor([0, 0, 0, 1, 1])

In [31]:
X_test = torch.tensor([
    [-0.8, 2.8],
    [2.6, -1.6],
])

y_test = torch.tensor([0, 1])

In [33]:
from torch.utils.data import Dataset


class ToyDataset(Dataset):
    def __init__(self, X, y):
        self.features = X
        self.labels = y

    def __getitem__(self, index):
        one_x = self.features[index]
        one_y = self.labels[index]        
        return one_x, one_y

    def __len__(self):
        return self.labels.shape[0]

train_ds = ToyDataset(X_train, y_train)
test_ds = ToyDataset(X_test, y_test)

In [34]:
print(train_ds[1])
print(len(train_ds))

(tensor([-0.9000,  2.9000]), tensor(0))
5


A custom PyTorch Dataset serves as a bridge between raw data and the training process, structured around three required methods:

* **`__init__` (Setup):** Initializes the data sources. While here it simply stores tensors in memory, it can also be used to define file paths or database connectors.
* **`__getitem__` (Fetch):** Defines how to retrieve a single record (feature and label) using an `index`. This is the method the `DataLoader` calls repeatedly to assemble a batch.
* **`__len__` (Count):** Returns the total number of items in the dataset (e.g., the number of rows in your label tensor). This tells the training loop when a full pass through the data (epoch) is complete.

The next step is passing these `Dataset` objects into a `DataLoader` to automate the batching and shuffling process.

In [35]:
from torch.utils.data import DataLoader

torch.manual_seed(123)

train_loader = DataLoader(
    dataset=train_ds,
    batch_size=2,
    shuffle=True,
    num_workers=0
)

In [36]:
test_ds = ToyDataset(X_test, y_test)

test_loader = DataLoader(
    dataset=test_ds,
    batch_size=2,
    shuffle=False,
    num_workers=0
)

In [37]:
for idx, (x, y) in enumerate(train_loader):
    print(f"Batch {idx+1}:", x, y)

Batch 1: tensor([[ 2.3000, -1.1000],
        [-0.9000,  2.9000]]) tensor([1, 0])
Batch 2: tensor([[-1.2000,  3.1000],
        [-0.5000,  2.6000]]) tensor([0, 0])
Batch 3: tensor([[ 2.7000, -1.5000]]) tensor([1])


Iterating through the `DataLoader` defines a *training epoch*, where every example in the dataset is seen exactly once. Here are some of the main features of DataLoader:

* **Dynamic Shuffling:** While a manual seed ensures initial reproducibility, the `DataLoader` reshuffles data differently in each subsequent epoch. This prevents repetitive update cycles, helping the model generalize better.
* **Handling Incomplete Batches:** If the dataset size isn't perfectly divisible by the `batch_size`, the final batch will be smaller (e.g., a batch of 1 when the size is 2). To prevent these small, "noisy" batches from destabilizing training convergence, it is best practice to drop the final incomplete batch using the `drop_last=True` parameter.

In [38]:
train_loader = DataLoader(
    dataset=train_ds,
    batch_size=2,
    shuffle=True,
    num_workers=0,
    drop_last=True,
)

In [40]:
for idx, (x, y) in enumerate(train_loader):
    print(f"Batch {idx+1}:", x, y)

Batch 1: tensor([[-0.9000,  2.9000],
        [ 2.3000, -1.1000]]) tensor([0, 1])
Batch 2: tensor([[ 2.7000, -1.5000],
        [-0.5000,  2.6000]]) tensor([1, 0])


The `num_workers` parameter in the `DataLoader` controls how many CPU processes are dedicated to loading and preprocessing data in the background. Wehn we set `num_workers=0`, data loading happens in the main process. This creates a bottleneck where the GPU sits idle because it must wait for the CPU to finish fetching the next batch before it can resume training. However, when `num_workers > 0`, multiple worker processes load data in parallel, pre-fetching batches so they are ready the moment the GPU finishes its current task.

<figure style='text-align: center'>
<img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/appendix-a_compressed/11.webp" width="800px">
</figure>

For tiny datasets or toy examples, setting `num_workers > 0` can actually slow things down due to the overhead of starting new processes. In interactive environments like Jupyter Notebooks, high worker counts can occasionally cause resource-sharing errors or crashes. While the ideal number depends on your CPU cores and hardware, `num_workers=4` is a common starting point for optimal performance on real-world datasets.

## **7. A typical training loop**



In [41]:
list(train_loader)

[[tensor([[ 2.3000, -1.1000],
          [ 2.7000, -1.5000]]),
  tensor([1, 1])],
 [tensor([[-1.2000,  3.1000],
          [-0.9000,  2.9000]]),
  tensor([0, 0])]]

In [73]:
import torch.nn.functional as F

torch.manual_seed(123)

model = NeuralNetwork(num_inputs=2, num_outputs=2)
optimizer = torch.optim.SGD(model.parameters(), lr=0.5)

num_epochs = 3
for epoch in range(num_epochs):

    model.train()
    for batch_idx, (features, labels) in enumerate(train_loader):
        
        logits = model(features)
        loss = F.cross_entropy(logits, target=labels)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        print(f"Epoch: {epoch+1:03d}/{num_epochs:03d} | "
              f"Batch: {batch_idx+1:03d}/{len(train_loader)} | "
              f"loss: {loss:.3f}"
        )

Epoch: 001/003 | Batch: 001/2 | loss: 0.749
Epoch: 001/003 | Batch: 002/2 | loss: 0.645
Epoch: 002/003 | Batch: 001/2 | loss: 0.442
Epoch: 002/003 | Batch: 002/2 | loss: 0.126
Epoch: 003/003 | Batch: 001/2 | loss: 0.027
Epoch: 003/003 | Batch: 002/2 | loss: 0.004


The code snippet above demonstrates a complete neural network optimization step, relying on specific configurations and practices to ensure stable convergence.

#### **Key Hyperparameters**
* **Learning Rate (`lr`):** A tunable setting that dictates the step size taken toward minimizing the loss.
* **Epochs:** The total number of complete passes through the training dataset.
* **Validation Set:** A separate dataset used repeatedly to fine-tune these hyperparameters without biasing the final evaluation on the test set.

#### **Training vs. Evaluation Modes**

`model.train()` and `model.eval()` toggle specific behaviors for layers like Dropout and Batch Normalization. Though optional for simple networks, including them prevents unexpected behavior if the architecture changes later.

#### **The Fused Loss & Backpropagation Loop**

* **`F.cross_entropy`:** Takes raw logits and applies the Softmax function internally for numerical efficiency and calculation stability.
* **`loss.backward()`**: Computes the gradients along the recorded computational graph.
* **`optimizer.step()`**: Executes the optimization algorithm (e.g., SGD), updating the weights by subtracting the scaled gradients.
* **`optimizer.zero_grad()`**: Clears the gradient history. Skipping this causes PyTorch to accumulate gradients across steps, leading to incorrect calculations and optimization failure.

After we trained the model, we can use it to make predictions, as shown below:

In [74]:
model.eval()

with torch.no_grad():
    outputs = model(X_train)

print(outputs)

torch.set_printoptions(sci_mode=False)
probas = torch.softmax(outputs, dim=1)
print(probas)

predictions = torch.argmax(probas, dim=1)
print(predictions)

tensor([[ 2.8569, -4.1618],
        [ 2.5382, -3.7548],
        [ 2.0944, -3.1820],
        [-1.4814,  1.4816],
        [-1.7176,  1.7342]])
tensor([[0.9991, 0.0009],
        [0.9982, 0.0018],
        [0.9949, 0.0051],
        [0.0491, 0.9509],
        [0.0307, 0.9693]])
tensor([0, 0, 0, 1, 1])


In [75]:
predictions = torch.argmax(outputs, dim=1)
print(predictions)

tensor([0, 0, 0, 1, 1])


In [76]:
predictions == y_train

tensor([True, True, True, True, True])

In [77]:
torch.sum(predictions == y_train)

tensor(5)

To generalize the computation of the prediction accuracy, let’s implement a compute_accuracy function as shown in the following code.

In [78]:
def compute_accuracy(model, dataloader):

    model.eval()
    correct = 0.0
    total_examples = 0
    
    for idx, (features, labels) in enumerate(dataloader):
        
        with torch.no_grad():
            logits = model(features)
        
        predictions = torch.argmax(logits, dim=1)
        compare = labels == predictions
        correct += torch.sum(compare)
        total_examples += len(compare)

    return (correct / total_examples).item()

In [79]:
compute_accuracy(model, train_loader)

1.0

The `compute_accuracy` function evaluates a model's performance by processing data through a `DataLoader` rather than loading the entire dataset at once. This approach has a few advantages as follows:

* **Memory Efficiency:** Large datasets cannot fit into GPU/CPU memory simultaneously. Iterating via a data loader ensures the model only processes a small, manageable chunk (batch) at any given moment.
* **Scalability:** This approach allows the evaluation metric to scale seamlessly to datasets of arbitrary size, using the same batch structure defined during training.
* **Consistency:** The internal logic mirrors standard inference, converting raw model logits into predicted class labels batch-by-batch to calculate the final accuracy score.

In [80]:
compute_accuracy(model, test_loader)

1.0

## **8. Saving and loading models**


PyTorch uses a model's `state_dict` to save and restore its learned parameters (weights and biases) efficiently. The recommended approach is to save only the parameters, not the entire model object. This creates a lightweight Python dictionary containing the model's weights. Standard practice for the file xtension is `.pth` or `.pt`.

In [84]:
torch.save(model.state_dict(), "model.pt")

To restore a saved model, we must recreate the empty architecture first and then inject the saved weights into it.

In [89]:
model = NeuralNetwork(2, 2) # needs to match the original model exactly
model.load_state_dict(torch.load("model.pt", weights_only=True))

<All keys matched successfully>

The code structure (`NeuralNetwork(2, 2)`) used during loading must match the original architecture exactly. If we add a layer or change a layer's size, PyTorch will throw a structural mismatch error because the saved weights won't line up with the new layers.

## **9. Optimizing training performance with GPUs**

See [code-part2.ipynb](code-part2.ipynb)